### PR TITLE
if_any() and if_all() to use inside filter()

### LINKED ISSUE #4770: Need equivalent of any_vars() with across()
Interim work around is to use `rowAny()`:

```R
rowAny <- function(x) rowSum(x) > 0
df %>% filter(rowAny(across(is.numeric, ~ .x > 0))
```

### LINKED ISSUE #5530: Example from `filter` doesn't work as suggested
In `filter_all `we read : 

```
# You can vary the selection of columns on which to apply the
# predicate. filter_at() takes a vars() specification:
filter_at(mtcars, vars(starts_with("d")), any_vars((. %% 2) == 0))
# ->
filter(mtcars, across(starts_with("d"), ~ (.x %% 2) == 0))
```

This is not right, as `any_vars()` doesn't have a direct replacement (this would be the replacement for `all_vars()`). 

```
library(dplyr, warn.conflicts = FALSE)
filter_at(mtcars, vars(starts_with("d")), any_vars((. %% 2) == 0))
#>                      mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4           21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag       21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
#> Datsun 710          22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
#> Hornet 4 Drive      21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
#> Hornet Sportabout   18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
#> Duster 360          14.3   8  360 245 3.21 3.570 15.84  0  0    3    4
#> Cadillac Fleetwood  10.4   8  472 205 2.93 5.250 17.98  0  0    3    4
#> Lincoln Continental 10.4   8  460 215 3.00 5.424 17.82  0  0    3    4
#> Chrysler Imperial   14.7   8  440 230 3.23 5.345 17.42  0  0    3    4
#> Dodge Challenger    15.5   8  318 150 2.76 3.520 16.87  0  0    3    2
#> AMC Javelin         15.2   8  304 150 3.15 3.435 17.30  0  0    3    2
#> Camaro Z28          13.3   8  350 245 3.73 3.840 15.41  0  0    3    4
#> Pontiac Firebird    19.2   8  400 175 3.08 3.845 17.05  0  0    3    2
filter(mtcars, across(starts_with("d"), ~ (.x %% 2) == 0))
#>  [1] mpg  cyl  disp hp   drat wt   qsec vs   am   gear carb
#> <0 lignes> (ou 'row.names' de longueur nulle)
```

### PR DESCRIPTION
**Pull request by romainfrancois**
``` r
library(palmerpenguins)
library(dplyr, warn.conflicts = FALSE)

penguins %>% 
  filter(if_any(starts_with("bill"), ~ . > 42))
#> # A tibble: 203 x 8
#>    species island bill_length_mm bill_depth_mm flipper_length_… body_mass_g
#>    <fct>   <fct>           <dbl>         <dbl>            <int>       <int>
#>  1 Adelie  Torge…           42.5          20.7              197        4500
#>  2 Adelie  Torge…           46            21.5              194        4200
#>  3 Adelie  Dream            42.2          18.5              180        3550
#>  4 Adelie  Dream            44.1          19.7              196        4400
#>  5 Adelie  Dream            42.3          21.2              191        4150
#>  6 Adelie  Torge…           45.8          18.9              197        4150
#>  7 Adelie  Torge…           42.8          18.5              195        4250
#>  8 Adelie  Torge…           42.1          19.1              195        4000
#>  9 Adelie  Torge…           42.9          17.6              196        4700
#> 10 Adelie  Dream            43.2          18.5              192        4100
#> # … with 193 more rows, and 2 more variables: sex <fct>, year <int>

penguins %>% 
  filter(if_all(starts_with("bill"), ~ . > 20))
#> # A tibble: 17 x 8
#>    species island bill_length_mm bill_depth_mm flipper_length_… body_mass_g
#>    <fct>   <fct>           <dbl>         <dbl>            <int>       <int>
#>  1 Adelie  Torge…           39.3          20.6              190        3650
#>  2 Adelie  Torge…           42            20.2              190        4250
#>  3 Adelie  Torge…           38.6          21.2              191        3800
#>  4 Adelie  Torge…           34.6          21.1              198        4400
#>  5 Adelie  Torge…           42.5          20.7              197        4500
#>  6 Adelie  Torge…           46            21.5              194        4200
#>  7 Adelie  Dream            39.2          21.1              196        4150
#>  8 Adelie  Dream            42.3          21.2              191        4150
#>  9 Adelie  Biscoe           41.3          21.1              195        4400
#> 10 Adelie  Dream            41.3          20.3              194        3550
#> 11 Adelie  Biscoe           45.6          20.3              191        4600
#> 12 Adelie  Biscoe           39.6          20.7              191        3900
#> 13 Adelie  Torge…           37.3          20.5              199        3775
#> 14 Adelie  Dream            40.2          20.1              200        3975
#> 15 Chinst… Dream            51.7          20.3              194        3775
#> 16 Chinst… Dream            54.2          20.8              201        4300
#> 17 Chinst… Dream            52            20.7              210        4800
#> # … with 2 more variables: sex <fct>, year <int>

# still works for now but warns
penguins %>% 
  filter(across(starts_with("bill"), ~ . > 20))
#> Warning: Using `across()` in `filter()` is deprecated, use `if_any()` or
#> `if_all()`
#> # A tibble: 17 x 8
#>    species island bill_length_mm bill_depth_mm flipper_length_… body_mass_g
#>    <fct>   <fct>           <dbl>         <dbl>            <int>       <int>
#>  1 Adelie  Torge…           39.3          20.6              190        3650
#>  2 Adelie  Torge…           42            20.2              190        4250
#>  3 Adelie  Torge…           38.6          21.2              191        3800
#>  4 Adelie  Torge…           34.6          21.1              198        4400
#>  5 Adelie  Torge…           42.5          20.7              197        4500
#>  6 Adelie  Torge…           46            21.5              194        4200
#>  7 Adelie  Dream            39.2          21.1              196        4150
#>  8 Adelie  Dream            42.3          21.2              191        4150
#>  9 Adelie  Biscoe           41.3          21.1              195        4400
#> 10 Adelie  Dream            41.3          20.3              194        3550
#> 11 Adelie  Biscoe           45.6          20.3              191        4600
#> 12 Adelie  Biscoe           39.6          20.7              191        3900
#> 13 Adelie  Torge…           37.3          20.5              199        3775
#> 14 Adelie  Dream            40.2          20.1              200        3975
#> 15 Chinst… Dream            51.7          20.3              194        3775
#> 16 Chinst… Dream            54.2          20.8              201        4300
#> 17 Chinst… Dream            52            20.7              210        4800
#> # … with 2 more variables: sex <fct>, year <int>
```

<sup>Created on 2020-12-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

closes #4770
closes #5530

### CODE REVIEW

**Comment by romainfrancois** [comment]
🤔 working on the tests, perhaps these names cause confusion when mixed with `tidyselect::all_of()` etc ...

``` r
library(palmerpenguins)
library(dplyr, warn.conflicts = FALSE)

bill <- c("bill_length_mm", "bill_depth_mm")
penguins %>% 
  filter(if_any(all_of(bill), ~ . > 42))
#> # A tibble: 203 x 8
#>    species island bill_length_mm bill_depth_mm flipper_length_… body_mass_g
#>    <fct>   <fct>           <dbl>         <dbl>            <int>       <int>
#>  1 Adelie  Torge…           42.5          20.7              197        4500
#>  2 Adelie  Torge…           46            21.5              194        4200
#>  3 Adelie  Dream            42.2          18.5              180        3550
#>  4 Adelie  Dream            44.1          19.7              196        4400
#>  5 Adelie  Dream            42.3          21.2              191        4150
#>  6 Adelie  Torge…           45.8          18.9              197        4150
#>  7 Adelie  Torge…           42.8          18.5              195        4250
#>  8 Adelie  Torge…           42.1          19.1              195        4000
#>  9 Adelie  Torge…           42.9          17.6              196        4700
#> 10 Adelie  Dream            43.2          18.5              192        4100
#> # … with 193 more rows, and 2 more variables: sex <fct>, year <int>

penguins %>% 
  filter(if_all(all_of(bill), ~ . > 20))
#> # A tibble: 17 x 8
#>    species island bill_length_mm bill_depth_mm flipper_length_… body_mass_g
#>    <fct>   <fct>           <dbl>         <dbl>            <int>       <int>
#>  1 Adelie  Torge…           39.3          20.6              190        3650
#>  2 Adelie  Torge…           42            20.2              190        4250
#>  3 Adelie  Torge…           38.6          21.2              191        3800
#>  4 Adelie  Torge…           34.6          21.1              198        4400
#>  5 Adelie  Torge…           42.5          20.7              197        4500
#>  6 Adelie  Torge…           46            21.5              194        4200
#>  7 Adelie  Dream            39.2          21.1              196        4150
#>  8 Adelie  Dream            42.3          21.2              191        4150
#>  9 Adelie  Biscoe           41.3          21.1              195        4400
#> 10 Adelie  Dream            41.3          20.3              194        3550
#> 11 Adelie  Biscoe           45.6          20.3              191        4600
#> 12 Adelie  Biscoe           39.6          20.7              191        3900
#> 13 Adelie  Torge…           37.3          20.5              199        3775
#> 14 Adelie  Dream            40.2          20.1              200        3975
#> 15 Chinst… Dream            51.7          20.3              194        3775
#> 16 Chinst… Dream            54.2          20.8              201        4300
#> 17 Chinst… Dream            52            20.7              210        4800
#> # … with 2 more variables: sex <fct>, year <int>
```

<sup>Created on 2020-12-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by romainfrancois** [comment]
Maybe also `if_all(everything())` and `if_any(everything())` is problematic ?

**Comment by romainfrancois** [comment]
This handles the combining internally based on an attribute of the data frame that is set by `if_any()` or `if_all()`, so that perhaps other functions could return data frames of logical vectors and decide about the combining. But this does not feel worth documenting it. 

This could be implemented separately from `across()` but given this is fairly esoteric, I felt this implementation would suffice. 

Perhaps later when we disallow data frame results columns, i.e. results straight from `across()` we can rewrite `if_any()` and `if_all()` so that they do the combining instead of letting the internals of `filter()` handle it.

File: NEWS.md [hunk 1]
```diff
@@ -19,6 +19,10 @@
 * `group_by()` uses the ungrouped data for the implicit mutate step (#5598). 
   You might have to define an `ungroup()` method for custom classes.
   For example, see https://github.com/hadley/cubelyr/pull/3. 
+  
+* New functions `if_any()` and `if_all()` to use inside `filter()` (#4770).
+
+* `group_by()` uses the ungrouped data for the implicit mutate step (#5598).
```
Comment (hadley) [inline NEWS.md:25]: I think you accidentally added this in a merge

File: tests/testthat/test-filter.r [hunk 2]
```diff
@@ -514,3 +493,16 @@ test_that("filter() preserves the call stack on error (#5308)", {
 
   expect_true(some(stack, is_call, "foobar"))
 })
+
+test_that("if_any() and if_all() work", {
```
Comment (hadley) [inline tests/testthat/test-filter.r:497]: Test that across in filter fails with an informative warning? (using a snapshot test)

File: vignettes/colwise.Rmd [hunk 3]
```diff
@@ -217,6 +211,34 @@ For some verbs, like `group_by()`, `count()` and `distinct()`, you can omit the
 
 `across()` doesn't work with `select()` or `rename()` because they already use tidy select syntax; if you want to transform column names with a function, you can use  `rename_with()`.
 
+### filter()
+
+We cannot directly use `across()` in `filter()` because we need an extra step to combine 
+the results. To that end, `filter()` has two special purpose companion functions: 
+
+* `if_any()` to keep the rows for which the predicate is true for *at least one* selected
```
Comment (hadley) [inline vignettes/colwise.Rmd:219]:
```suggestion
* `if_any()` keeps the rows where the predicate is true for *at least one* selected
```

File: vignettes/colwise.Rmd [hunk 4]
```diff
@@ -217,6 +211,34 @@ For some verbs, like `group_by()`, `count()` and `distinct()`, you can omit the
...
+  column: 
+  
+```{r}
+starwars %>% 
+  filter(if_any(everything(), ~ !is.na(.x)))
+```
+  
+* `if_all()` to keep the rows where the predicate is true for *all* selected columns:
```
Comment (hadley) [inline vignettes/colwise.Rmd:227]:
```suggestion
* `if_all()` keeps the rows where the predicate is true for *all* selected columns:
```

**Review by hadley — COMMENTED** [review]
I reviewed the API. It would be good to get a review on the C++ too (just in case)

File: tests/testthat/test-filter.r (hunk 2 above)
Comment (romainfrancois) [reply tests/testthat/test-filter.r:503]: added it to the existing `verify_output()`, will move to snapshot later.

File: src/filter.cpp [hunk 5]
```diff
@@ -111,13 +147,43 @@ SEXP eval_filter_one(SEXP quos, SEXP mask, SEXP caller, R_xlen_t n, SEXP env_fil
...
+          combine_and = false;
+        }
+      }
+      if (first && warn) {
+        SEXP expr = rlang::quo_get_expr(VECTOR_ELT(quos, i));
+        bool across = TYPEOF(expr) == LANGSXP && CAR(expr) == dplyr::symbols::across;
+        if (across) {
+          Rf_warningcall(R_NilValue, "Using `across()` in `filter()` is deprecated, use `if_any()` or `if_all()`");
+        } else {
+          Rf_warningcall(R_NilValue, "data frame results in `filter()` are deprecated, use `if_any()` or `if_all()`");
```
Comment (romainfrancois) [inline src/filter.cpp:169]: Perhaps this should drop the ", use if_any ..." bit as data frames here might come from some other construct,

**Comment by romainfrancois** [comment]
I've removed the warnings when using `across()` and data frame results for now, so that users can migrate `filter(across())` -> `filter(if_any())` silently for now. 

`filter(across())` does the same thing as `filter(if_all())` without warning for this 1.0.4 release, and then later we can start to warn.

**Pull request merged**